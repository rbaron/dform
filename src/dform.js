
const shouldRender = (state, cond) => {
  if (cond.type === 'always_true') {
    return true
  } else if (cond.type === 'equals') {
    return state[cond.field_id] === cond.field_value;
  } else if (cond.type === 'not_empty') {
    return state[cond.field_id] !== '';
  } else if (cond.type === 'truthy') {
    return !!state[cond.field_id];
  } else if (cond.type === 'falsy') {
    return !state[cond.field_id];
  } else if (cond.type === 'and') {
    return cond.conds.every(c => shouldRender(state, c))
  } else if (cond.type === 'or') {
    return cond.conds.some(c => shouldRender(state, c))
  } else {
    return false
  }
}

const renderInput = (field, inputFactories, missingFactoryHandler) => {
  if (field.type in inputFactories) {
    return inputFactories[field.type](field);
  } else if (missingFactoryHandler) {
    return missingFactoryHandler(field);
  } else {
    throw new Error(`No input factory found for ${field.type}`);
  }
}

const renderForm = (state, schema, inputFactories, missingFactoryHandler = null) => {
  if (shouldRender(state, schema.cond)) {
    const fields = schema.fields.map(f => renderInput(f, inputFactories, missingFactoryHandler))

    const children = schema.children ? schema.children.reduce((acc, c) =>
      [ ...acc, ...renderForm(state, c, inputFactories, missingFactoryHandler)], []) : []

    return [
      ...fields,
      ...children,
    ]
  } else {
    return [];
  }
}

const activeFields = (state, schema) => {
  const inputFactories = {}
  const missingFactoryHandler = field => field

  return renderForm(state, schema, inputFactories, missingFactoryHandler)
}

const defaultState = (schema, keyExtractor) => {
  const getDefault = field => {
    switch (field.type) {
      case 'boolean':
        return field.default || false
      case 'string':
        return field.default || ''
      default:
        return field.default
    }
  }

  const traverse = (schema) => {
    const state = schema.fields.reduce((acc, f) => {
      return { ...acc, [keyExtractor(f)]: getDefault(f) }
    }, {})

    if (!schema.children) {
      return state
    } else {
      return schema.children.reduce((acc, child) => ({
        ...acc,
        ...traverse(child),
      }), state)
    }
  }

  return traverse(schema)
}

export { activeFields, defaultState, renderForm }
