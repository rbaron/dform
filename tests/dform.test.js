import { activeFields, renderForm } from '../src/dform'

const schema = {
  // height 1 node 1
  fields: [{
    type: 'string',
    id: 'field_111'
  }, {
    type: 'boolean',
    id: 'field_112'
  }],
  cond: {
    type: 'always_true',
  },
  children: [{
  // height 2 node 1
    fields: [{
      type: 'boolean',
      id: 'field_211',
    }, {
      type: 'string',
      id: 'field_212',
    }],
    cond: {
      type: 'and',
      conds: [{
        type: 'not_empty',
        field_id: 'field_111',
      }, {
        type: 'truthy',
        field_id: 'field_112',
      }],
    },
    children: [{
      // height 3 node 1
      fields: [{
        type: 'string',
        id: 'field_311',
      }],
      cond: {
        type: 'always_true',
      },
    }],
  }, {
    // height 2 node 2
    fields: [{
      type: 'boolean',
      id: 'field_221',
    }, {
      type: 'string',
      id: 'field_222',
    }],
    cond: {
      type: 'or',
      conds: [{
        type: 'not_empty',
        field_id: 'field_111',
      }, {
        type: 'truthy',
        field_id: 'field_112',
      }],
    },
    children: [{
      // height 3 node 2
      fields: [{
        type: 'string',
        id: 'field_321',
      }],
      cond: {
        type: 'equals',
        field_id: 'field_221',
        field_value: true,
      },
    }],
  }],
}

const inputFactories = {
  'string': (args) => `<StringInput ${JSON.stringify(args)}>`,
  'boolean': (args) => `<BooleanInput ${JSON.stringify(args)}>`,
}

describe('render', () => {
  it('renders all fields', () => {
    const state = {
      field_111: 'Some content',
      field_112: true,
      field_221: true,
    }

    const rendered = renderForm(state, schema, inputFactories);
    expect(rendered).toHaveLength(8)
    expect(rendered).toMatchSnapshot()
  })

  it('renders node 11 and node 22 fields', () => {
    const state = {
      field_111: '',
      field_112: true,
    }

    const rendered = renderForm(state, schema, inputFactories);
    expect(rendered).toHaveLength(4)
    expect(rendered).toMatchSnapshot()
  })

  it('renders node 11, node 22 and node 32 fields', () => {
    const state = {
      field_111: '',
      field_112: true,
      field_221: true,
    }

    const rendered = renderForm(state, schema, inputFactories);
    expect(rendered).toHaveLength(5)
    expect(rendered).toMatchSnapshot()
  })

  it('renders node 11 and node 22 fields', () => {
    const state = {
      field_111: '',
      field_112: true,
      field_221: false,
    }

    const rendered = renderForm(state, schema, inputFactories);
    expect(rendered).toHaveLength(4)
    expect(rendered).toMatchSnapshot()
  })
})

describe('activeKeys', () => {
  it('returns all fields', () => {
    const state = {
      field_111: 'Some content',
      field_112: true,
      field_221: true,
    }

    const fields = activeFields(state, schema)
    const fieldSet = new Set(fields.map(f => f.id))

    const expected = new Set([
      'field_111',
      'field_112',
      'field_211',
      'field_212',
      'field_311',
      'field_221',
      'field_222',
      'field_321',
    ])
    expect(fieldSet).toEqual(expected)
  })

  it.only('returns expected fields', () => {
    const state = {
      field_111: 'Some content',
      field_112: false,
      field_221: true,
    }
    const fields = activeFields(state, schema)
    const fieldSet = new Set(fields.map(f => f.id))

    const expected = new Set([
      'field_111',
      'field_112',
      'field_221',
      'field_222',
      'field_321',
    ])
    expect(fieldSet).toEqual(expected)
  })
})
