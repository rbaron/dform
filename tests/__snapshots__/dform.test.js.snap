// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders all fields 1`] = `
Array [
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_111\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_112\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_211\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_212\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_311\\",\\"default\\":\\"field_311 default value\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_221\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_222\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_321\\"}>",
]
`;

exports[`render renders node 11 and node 22 fields 1`] = `
Array [
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_111\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_112\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_221\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_222\\"}>",
]
`;

exports[`render renders node 11 and node 22 fields 2`] = `
Array [
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_111\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_112\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_221\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_222\\"}>",
]
`;

exports[`render renders node 11, node 22 and node 32 fields 1`] = `
Array [
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_111\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_112\\"}>",
  "<BooleanInput {\\"type\\":\\"boolean\\",\\"id\\":\\"field_221\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_222\\"}>",
  "<StringInput {\\"type\\":\\"string\\",\\"id\\":\\"field_321\\"}>",
]
`;
